### PR TITLE
Move mono repo config to nx targets rather than npm scripts based

### PR DESCRIPTION
**Pull request by Roaders**
This PR was initially supposed to just be a case of moving npm scripts to nx targets. It ended up being much bigger than that though:

- move npm scripts to nx targets
- move from webpack to vite for test-harness (the nx webpack implementation did not allow us to do what we wanted for the dev server)
- move to ESM (there were issues with vite that made it easier to move to vite now. We would have done this anyway but it seemed easier to do it as part of this PR. We can have a discussion at some point about dual publish)
- move to vitest (same reasons as above but @paulgoldsmith did this work in another PR. I have to merge this in though)
as a result I think I probably ended up touching nearly every file in the repo. Apologies for the huge PR!

### CODE REVIEW

File: projects/lib/src/agent/desktop-agent.spec.ts [hunk 1]
```diff
@@ -29,23 +29,23 @@ import {
     setupProperty,
 } from '@morgan-stanley/ts-mocking-bird';
 import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';
-import { peerDependencies } from '../../package.json';
-import { AppDirectory } from '../app-directory';
-import { AppDirectoryApplication } from '../app-directory.contracts';
-import { ChannelFactory, Channels } from '../channel';
-import { ChannelMessageHandler } from '../channel/channel-message-handler';
-import { HEARTBEAT } from '../constants';
+import packageJson from '../../package.json';
```
[the remark this answers is not in the document]
Comment (Roaders) [reply projects/lib/src/agent/desktop-agent.spec.ts:32]: it wasn't happy with a destructuring of a json file after the conversion to esm.

File: projects/lib/src/channel/channels.spec.ts [hunk 2]
```diff
@@ -34,12 +34,12 @@ import {
...
     ResponseMessage,
-} from '../contracts';
-import { isGetCurrentChannelRequest } from '../helpers';
-import * as helpersImport from '../helpers';
-import { ContextListener } from './channel.contracts';
-import { Channels } from './channels';
-import { ChannelFactory } from './channels.factory';
+} from '../contracts.js';
+import * as helpersImport from '../helpers/index.js';
+import { isGetCurrentChannelRequest } from '../helpers/index.js';
```
[the remark this answers is not in the document]
Comment (Roaders) [reply projects/lib/src/channel/channels.spec.ts:39]: the tests pass so I presume so... A good point though. I guess they resolve to the same place so vi is ok with it

**Comment by paulgoldsmith** [comment]
There is no coverage summary in the build output now. Is it possible to enable that?

**Comment by paulgoldsmith** [comment]
CodeCov has gone missing

{16303C83-2CAD-4888-93B3-5C086725622F}

**Comment by Roaders** [comment]
> There is no coverage summary in the build output now. Is it possible to enable that?

This was the next thing that I was going to check. Coverage is still generated but hadn't checked what we did with it.

File: projects/lib/src/channel/channels.spec.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (Roaders) [reply projects/lib/src/channel/channels.spec.ts:39]: I've updated the paths

**Comment by Roaders** [comment]
> > > There is no coverage summary in the build output now. Is it possible to enable that?
> > 
> > 
> > This was the next thing that I was going to check. Coverage is still generated but hadn't checked what we did with it.
> 
> I'm not seeing coverage summary in the build output. CodeCov is nice, though I like the vitest command line summary too for reviewing. Not critical cause I can run `npx vitest --coverage` locally, but if it is easy to switch on it'd be good.


Is this what you mean:

image

**Comment by Roaders** [comment]
I think that the coverage summary for fdc3-web might be cutoff because of the amount of console-log that there is...

**Comment by paulgoldsmith** [comment]
> I think that the coverage summary for fdc3-web might be cutoff because of the amount of console-log that there is...

Yes, that could be it. Do you see it locally when you run `build-release`?

**Comment by paulgoldsmith** [comment]
I pulled the PR locally and determined the way to view all test output and coverage is to run `npx nx run-many -t test --coverage --verbose`. Running using the npm script is not possible as it cannot pass down the `coverage` and `verbose` arguments through to Nx Vite plugin. Can we introduce another couple of npm scripts to run locally to see the coverage? Like the following

```
 "test:coverage": "nx run-many -t test --skip-nx-cache -- --coverage --verbose"
 "test:coverage:watch": "nx run-many -t test --skip-nx-cache -- --coverage --verbose --watch"
```

**Comment by Roaders** [comment]
the `--coverage` flag is not needed as coverage is enabled by default. Personally I don't think that we need to see the coverage stats when running tests across all projects like this - as long as it reaches the minimum level we're fine (although I have just seen that we don't seem to be setting minimums so I'll add that).

I also can't really see the benefit of running the watch command. I would just run the watch test command for whatever project I am working on:

```
nx test fdc3-web --watch
```

I don't understand how nx would run multiple targets with `--watch`. Presumably if there were more than 3 targets (or whatever the minimum parallel threads is set to) then some of the targets would never run as they would be waiting for the previous ones to finish.

**Comment by Roaders** [comment]
I have updated the readme with examples of how you run tests on individual projects.

**Comment by Roaders** [comment]
...also added thresholds so if the coverage gets below 80 the build will fail

**Comment by Roaders** [comment]
I've also added an `npm run test:watch` command that will run and watch tests across all projects. This runs a single target in the root that will include all files.

**Review by paulgoldsmith — APPROVED** [review]
